Similarly named method signature clash resulting in compile time error

The following program `yaegi_wrong_method.go` triggers a **compile error**:

```go
package main

import "fmt"

type WidgetEvent struct {
        Nothing string
}

type WidgetControl interface {
        HandleEvent(e *WidgetEvent)
}

type Button struct{}

func (b *Button) HandleEvent(e *WidgetEvent) {
}

type WindowEvent struct {
        Something int
}

type Window struct {
        Widget WidgetControl
}

func (w *Window) HandleEvent(e *WindowEvent) {
}

func main() {
        window := Window{
                Widget: &Button{},
        }
        windowevent := &WindowEvent{}
        // The next line uses the signature from the wrong method, resulting in an error.
        // Renaming one of the clashing method names fixes the problem.
        window.HandleEvent(windowevent)
        fmt.Println("OK!")
}
```

Expected result:
```console
$ go run ./yaegi_wrong_method.go
OK!
```

Got:
```console
$ yaegi ./yaegi_wrong_method.go
run: yaegi_wrong_method.go:36:21: cannot use type *main.WindowEvent as type
```

Error started from commit: e32b2ab6bdfc16f225868c979e565eb0340340ff

Workaround: change the name of one of the clashing methods.

I've noticed that if I change the test code slightly to call the method on a _reference_ of type `Window`...

```
...
window := &Window{
...
```

...I get a new error (not seen using `go run`)...

```
> yaegi ./yaegi_wrong_method.go
run: ./yaegi_wrong_method.go:36:21: cannot use type *main.WindowEvent as type
```

----------

Also, in my application code I get a different type of error - I'll see if I can work out the small repro for this but the error is of the form...

```
reflect.Value.Convert: value of type *game.Input cannot be converted to type *struct { Type int; PosArea image.Point; Po
sPixels image.Point; WheelDelta image.Point; Keys []int32 }
```

...which, given my apps code, is still comparing the wrong types from methods of the same name somehow.

--- a/_test/issue-1163.go
+++ b/_test/issue-1163.go
@@ -27,7 +27,7 @@ func (w *Window) HandleEvent(e *WindowEvent) {
 }
 
 func main() {
-	window := Window{
+	window := &Window{
 		Widget: &Button{},
 	}
 	windowevent := &WindowEvent{}

--- a/_test/issue-1163.go
+++ b/_test/issue-1163.go
@@ -1,0 +1,41 @@
+package main
+
+import "fmt"
+
+type WidgetEvent struct {
+	Nothing string
+}
+
+type WidgetControl interface {
+	HandleEvent(e *WidgetEvent)
+}
+
+type Button struct{}
+
+func (b *Button) HandleEvent(e *WidgetEvent) {
+}
+
+type WindowEvent struct {
+	Something int
+}
+
+type Window struct {
+	Widget WidgetControl
+}
+
+func (w *Window) HandleEvent(e *WindowEvent) {
+}
+
+func main() {
+	window := Window{
+		Widget: &Button{},
+	}
+	windowevent := &WindowEvent{}
+	// The next line uses the signature from the wrong method, resulting in an error.
+	// Renaming one of the clashing method names fixes the problem.
+	window.HandleEvent(windowevent)
+	fmt.Println("OK!")
+}
+
+// Output:
+// OK!

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1525,6 +1525,18 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 							rtype := n.typ.TypeOf()
 							n.typ = &itype{cat: valueT, rtype: rtype, val: n.typ}
 						}
+					case isStruct(n.typ):
+						// If a method of the same name exists, use it if it is less shallowed.
+						// if method's depth is the same as field's, this is an error.
+						d := n.typ.methodDepth(n.child[1].ident)
+						if d >= 0 && d < len(ti) {
+							goto tryMethods
+						}
+						if d == len(ti) {
+							err = n.cfgErrorf("ambiguous selector: %s", n.child[1].ident)
+							break
+						}
+						fallthrough
 					default:
 						n.gen = getIndexSeq
 						n.typ = n.typ.fieldSeq(ti)
@@ -1540,11 +1552,24 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 					// Handle an embedded binary field into a struct field.
 					n.gen = getIndexSeqField
 					lind = append(lind, s.Index...)
+					if isStruct(n.typ) {
+						// If a method of the same name exists, use it if it is less shallowed.
+						// if method's depth is the same as field's, this is an error.
+						d := n.typ.methodDepth(n.child[1].ident)
+						if d >= 0 && d < len(lind) {
+							goto tryMethods
+						}
+						if d == len(lind) {
+							err = n.cfgErrorf("ambiguous selector: %s", n.child[1].ident)
+							break
+						}
+					}
 					n.val = lind
 					n.typ = &itype{cat: valueT, rtype: s.Type}
 					break
 				}
 				// No field (embedded or not) matched. Try to match a method.
+			tryMethods:
 				fallthrough
 			default:
 				// Find a matching method.

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1553,7 +1553,7 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 					n.gen = getIndexSeqField
 					lind = append(lind, s.Index...)
 					if isStruct(n.typ) {
-						// If a method of the same name exists, use it if it is less shallowed.
+						// If a method of the same name exists, use it if it is shallower than the struct field.
 						// if method's depth is the same as field's, this is an error.
 						d := n.typ.methodDepth(n.child[1].ident)
 						if d >= 0 && d < len(lind) {

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1512,7 +1512,7 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 				// Find a matching field.
 				if ti := n.typ.lookupField(n.child[1].ident); len(ti) > 0 {
 					if isStruct(n.typ) {
-						// If a method of the same name exists, use it if it is less shallowed.
+						// If a method of the same name exists, use it if it is shallower than the struct field.
 						// if method's depth is the same as field's, this is an error.
 						d := n.typ.methodDepth(n.child[1].ident)
 						if d >= 0 && d < len(ti) {

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -1511,6 +1511,18 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 			case isStruct(n.typ) || isInterfaceSrc(n.typ):
 				// Find a matching field.
 				if ti := n.typ.lookupField(n.child[1].ident); len(ti) > 0 {
+					if isStruct(n.typ) {
+						// If a method of the same name exists, use it if it is less shallowed.
+						// if method's depth is the same as field's, this is an error.
+						d := n.typ.methodDepth(n.child[1].ident)
+						if d >= 0 && d < len(ti) {
+							goto tryMethods
+						}
+						if d == len(ti) {
+							err = n.cfgErrorf("ambiguous selector: %s", n.child[1].ident)
+							break
+						}
+					}
 					n.val = ti
 					switch {
 					case isInterfaceSrc(n.typ):
@@ -1525,18 +1537,6 @@ func (interp *Interpreter) cfg(root *node, importPath string) ([]*node, error) {
 							rtype := n.typ.TypeOf()
 							n.typ = &itype{cat: valueT, rtype: rtype, val: n.typ}
 						}
-					case isStruct(n.typ):
-						// If a method of the same name exists, use it if it is less shallowed.
-						// if method's depth is the same as field's, this is an error.
-						d := n.typ.methodDepth(n.child[1].ident)
-						if d >= 0 && d < len(ti) {
-							goto tryMethods
-						}
-						if d == len(ti) {
-							err = n.cfgErrorf("ambiguous selector: %s", n.child[1].ident)
-							break
-						}
-						fallthrough
 					default:
 						n.gen = getIndexSeq
 						n.typ = n.typ.fieldSeq(ti)

--- a/interp/type.go
+++ b/interp/type.go
@@ -1372,6 +1372,17 @@ func (t *itype) lookupMethod(name string) (*node, []int) {
 	return m, index
 }
 
+// methodDepth returns a depth greater or equal to 0, or -1 if no match.
+func (t *itype) methodDepth(name string) int {
+	if m, lint := t.lookupMethod(name); m != nil {
+		return len(lint)
+	}
+	if _, lint, _, ok := t.lookupBinMethod(name); ok {
+		return len(lint)
+	}
+	return -1
+}
+
 // LookupBinMethod returns a method and a path to access a field in a struct object (the receiver).
 func (t *itype) lookupBinMethod(name string) (m reflect.Method, index []int, isPtr, ok bool) {
 	if t.cat == ptrT {

